types/wlr_pointer_constraints_v1.c

```diff
@@ -0,0 +1,387 @@
+static void send_constrained(struct wlr_pointer_constraint_v1 *constraint) {
+	constraint->active = true;
+
+	wlr_log(L_DEBUG, "constrained %p", constraint);
+	zwp_confined_pointer_v1_send_confined(constraint->resource);
+}
+static void send_unconstrained(struct wlr_pointer_constraint_v1 *constraint) {
+	if (!constraint->active) {
+		return;
+	}
+	constraint->active = false;
+
+	wlr_signal_emit_safe(
+		&constraint->pointer_constraints->events.constraint_inactive,
+		constraint);
+
+	wlr_log(L_DEBUG, "unconstrained %p", constraint);
+	zwp_confined_pointer_v1_send_unconfined(constraint->resource);
+	if (constraint->lifetime == ZWP_POINTER_CONSTRAINTS_V1_LIFETIME_ONESHOT) {
+		wl_resource_destroy(constraint->resource);
+	}
+}
+
+static void pointer_constraint_check_active(struct wlr_pointer_constraint_v1 *constraint) {
+	struct wlr_pointer_constraint_v1 *current_constraint =
+		constraint->seat_client->seat->current_pointer_constraint;
+	if (current_constraint != constraint ||
+			constraint->current.region || !constraint->accepted) {
+		return;
+	}
+
+	struct wlr_pointer_constraint_v1_activation activation = {
+		.pointer_constraint = constraint,
+		.box = NULL,
+	};
+
+	send_constrained(constraint);
+	wlr_signal_emit_safe(
+		&constraint->pointer_constraints->events.constraint_active,
+		&activation);
+}
+
+void wlr_pointer_constraint_v1_accept(struct wlr_pointer_constraint_v1 *constraint) {
+	wlr_log(L_DEBUG, "constraint %p accepted", constraint);
+	if (constraint->accepted) {
+		return;
+	}
+	constraint->accepted = true;
+	pointer_constraint_check_active(constraint);
+}
+
+void wlr_pointer_constraint_v1_decline(struct wlr_pointer_constraint_v1 *constraint) {
+	wlr_log(L_DEBUG, "constraint %p declined", constraint);
+	constraint->accepted = false;
+	send_unconstrained(constraint);
+}
+
+static void resource_destroy(struct wl_client *client,
+		struct wl_resource *resource) {
+	wl_resource_destroy(resource);
+}
+
+static void pointer_constraint_destroy(struct wlr_pointer_constraint_v1 *constraint) {
+	if (constraint == NULL) {
+		return;
+	}
+
+	if (constraint->seat_client->seat->current_pointer_constraint == constraint) {
+		constraint->seat_client->seat->current_pointer_constraint = NULL;
+	}
+
+	send_unconstrained(constraint);
+	wl_resource_set_user_data(constraint->resource, NULL);
+	wl_list_remove(&constraint->link);
+	wl_list_remove(&constraint->surface_commit.link);
+	wl_list_remove(&constraint->surface_destroy.link);
+	free(constraint);
+}
+
+static void pointer_constraint_destroy_resource(struct wl_resource *resource) {
+	struct wlr_pointer_constraint_v1 *constraint =
+		pointer_constraint_from_resource(resource);
+
+	pointer_constraint_destroy(constraint);
+}
+
+static void pointer_constraint_set_region(struct wl_client *client,
+		struct wl_resource *resource, struct wl_resource *region) {
+	struct wlr_pointer_constraint_v1 *constraint =
+		pointer_constraint_from_resource(resource);
+
+	constraint->pending.region = region;
```

Comment:
The protocl says you have to handle a NULL region here by using the input region of the surface.

Reply:
I do handle that, just not here obviously.